File: 0701/SeoyeonHong/합승 택시 요금.py
# https://school.programmers.co.kr/learn/courses/30/lessons/72413
import heapq

def solution(n, s, a, b, fares):
    graph = [[] for _ in range(n+1)] # 인접 리스트
    INF = int(1e9)
    
    for c, d, f in fares:
        graph[c].append((d, f))
        graph[d].append((c, f))
    
    def dijkstra(start, fare):
        q = [(0, start)]
        fare[start] = 0
        while q:
            f, d = heapq.heappop(q)
            if fare[d] >= f:
                for nd, nf in graph[d]:
                    if f + nf < fare[nd]:
                        fare[nd] = f + nf
                        heapq.heappush(q, (f + nf, nd))
    
    toM = [INF for _ in range(n+1)]
    dijkstra(s, toM) # s번 지점으로부터 m번 지점까지의 최소 비용 계산(합승 요금)
    
    toA, toB = [INF for _ in range(n+1)], [INF for _ in range(n+1)]
    dijkstra(a, toA) # a번 지점에서 m번 지점까지의 최소 비용 계산(개인 요금)
    dijkstra(b, toB) # b번 지점에서 m번 지점까지의 최소 비용 계산(개인 요금)
    
        
    totalFares = [INF for _ in range(n+1)] # m번 지점까지 합승했을 때의 총 요금
    for m in range(1, n+1):
        totalFares[m] = toM[m] + toA[m] + toB[m]
                
    return min(totalFares)

File: 0701/SoohyunKim/합승택시요금.js
// https://school.programmers.co.kr/learn/courses/30/lessons/72413

function solution(n, s, a, b, fares) {
    const board = Array.from({ length: n }, () => Array(n).fill(Infinity));
    for(let i = 0; i < n; i++) {
        board[i][i] = 0;
    }
    
    fares.forEach(el => {
        const [x, y, weight] = el;
        board[x-1][y-1] = weight;
        board[y-1][x-1] = weight;
    });
  
    // Floyd-Warshall
    for(let k = 0; k < n; k++) { // 경유
        for(let i = 0; i < n; i++) { // 출발
            for(let j = 0; j < n; j++) { // 도착
                if (distance[i][j] > distance[i][k] + distance[k][j]) {
                    distance[i][j] = distance[i][k] + distance[k][j];
                }
            }
        }
    }
  
    // 합승X
    let answer = board[s-1][a-1] + board[s-1][b-1];

    for(let i = 0; i < n; i++) {
        // 합승O
        const shortest = board[s-1][i] + board[i][a-1] + board[i][b-1]; 
        answer = Math.min(answer, shortest);
    }
  
    return answer;
}

